src/citrine/resources/file_link.py

```diff
@@ -149,6 +149,31 @@ def __init__(self, project_id: UUID, dataset_id: UUID, session: Session):
         self.dataset_id = dataset_id
         self.session = session
 
+    def _get_versioned_path(self,
+                            file_id: Union[str, UUID],
+                            version_id: Union[str, UUID],
+                            *,
+                            action: str = None) -> str:
+        """Build the path for taking an action with a particular file version."""
+        base = urlparse(self._get_path(file_id))
+        new_path = base.path.split('/')
+        new_path.extend([quote(x) for x in ('versions', str(version_id))])
+        if action is not None:
+            new_path.append(quote(action))
+
+        return base._replace(path='/'.join(new_path)).geturl()
+
+    def _get_path_from_file_link(self, file_link: FileLink,
+                                 *,
+                                 action: str = None) -> str:
+        """Build the path for taking an action with a particular file link."""
+        # Use this sessions project/dataset credentials and the URL's file / version
+        parsed = urlparse(file_link.url)
```

Comment:
Shouldn't the file link object itself just store the file and version IDs are properties? Actually seems like this whole thing should live in FileLink, no?

Reply:
We should arguably be storing the versioned_url returned by the `files` routes, but the S3 resolution route on the backend literally expects a relative URL of the structure `/....../dataset/{uuid}/file/{uuid}/project/{uuid}`.  I'd like to move this full protocol over to where FileLinks are only storing those two pieces of information for platform files.  That will need to be a multi-PR multi-library strategy because we don't want to break anything that exists today.  This PR makes citrine-python flexible in what it consumes while still emitting what other tool expect.

[This parse](https://github.com/CitrineInformatics/platform-backend/blob/1cd99ce0cee1acc652d31027f400268f67305811/code/rest/clients/src/main/scala/io/citrine/rest/clients/data/api/impl/akka/FileServiceApiImpl.scala#L84-L88) feeds [this construction](https://github.com/CitrineInformatics/platform-backend/blob/1cd99ce0cee1acc652d31027f400268f67305811/code/rest/clients/src/main/scala/io/citrine/rest/clients/data/api/impl/akka/FileServiceApiImpl.scala#L70).  Note that nothing is done with datasetID even though the parse requires it.  If we got rid of that, we could just store that URL.